### DISCUSSION TITLE
Handling Network Issues in Remix

### DISCUSSION THREAD

**Discussion by rubber-duck-software** [remix-run/remix#5955] [Q&A] [answered]
I’m working on a Remix app where gracefully handling network issues is a high priority. I’ve recently run into trouble with error handling in Remix and was wondering how to handle this. 

**The Error Handling I’m after is this:**
If a form submission fails due to network issues, we show the user a dialog which informs them of the failure. When the dialog is dismissed, the user should be seeing their form again still filled out and ready to retry the submission. (I think this is pretty typical error handling for mutations in an SPA)

**Summary of the problem:**
If an network issue occurs causing us to never receive a response from the action function, Remix will render the route’s ErrorBoundary causing us to lose whatever data was in the Form. 

**Additional Information**
I have an example repro (https://github.com/rubber-duck-software/remix-error-handling-example) which gives more context and a corresponding site which demonstrates the issue (https://jade-fudge-07dcb3.netlify.app). 

I have a workaround which achieves the intended behavior (see example repo and example site), but it cannot work across a whole app (involves ugly routing structure, adding path segments, and other nonsense).

**Comment by kiliman** [comment, marked as answer]
The link to your repo returns 404.

Anyway, yes, that appears to be a limitation of Remix. The issue is that Remix will render your `ErrorBoundary` on any error, and when it does, it replaces your UI component with the boundary, thereby losing its state.

It would be nice if you could embed the `ErrorBoundary` inside your UI component or perhaps an `ErrorOutlet` where you can specify where the ErrorBoundary should render.

I think the reason why they replace your UI component is that there's no way to tell what the state of your UI component is. It could be that simply rendering your UI could throw an error, and if it tried to re-render your UI component as well, you'd get caught in an infinite loop.

Anyway, I do think there needs to be an escape hatch where you can tell Remix, "let me manually handle errors. If I can't handle it, I'll rethrow back to you."

**Reply by rubber-duck-software** [reply]
Hi @kiliman - Thanks for your quick response. I've opened up a proposal here https://github.com/remix-run/remix/discussions/5957 . Also apologies- forgot to make that repo public. Should be fixed now. 

As for the ErrorBoundary behavior. I think Remix's behavior is correct for errors in the loader or when rendering the component. (You make a good point about infinite loops). 

But when the error is raised on a mutation, the UI component will always be able to continue rendering. For this reason I think action errors are a bit of a special case.
